File: ASM/hw/hw2/main.cpp
#include <iostream>
#include <string>
#include <vector>
#include <cctype>

namespace Lexem {
    enum Type {
        UNKNOWN,
        TYPE,
        ID,
        LPAREN,
        RPAREN,
        LBRACKET,
        RBRACKET,
        STAR,
        COMMA,
        SEMICOLON,
        NUMBER,
        COUNT
    };
}

namespace State {
    enum Type {
        ERROR,
        START,
        TYPE,
        ID,
        LPAREN,
        PARAM_LIST, // Переименовано PARAMS -> PARAM_LIST для ясности
        PARAM_STAR,
        ARRAY_DIM, // Переименовано NUMBER -> ARRAY_DIM для ясности
        NUMBER,
        RPAREN,
        END,
        COUNT
    };
}

// Таблица переходов автомата (исправлена)
static const State::Type transitionTable[State::COUNT][Lexem::COUNT] = {
        //  UNKNOWN  TYPE    ID      LPAREN  RPAREN  LBRACKET  RBRACKET  STAR  COMMA  SEMICOLON  NUMBER
        /* ERROR */      {State::ERROR,      State::ERROR,      State::ERROR,      State::ERROR,      State::ERROR,      State::ERROR,   State::ERROR,   State::ERROR,      State::ERROR,   State::ERROR,      State::ERROR},
        /* START */      {State::ERROR,      State::TYPE,       State::ERROR,      State::ERROR,      State::ERROR,      State::ERROR,   State::ERROR,   State::ERROR,      State::ERROR,   State::ERROR,      State::ERROR},
        /* TYPE */       {State::ERROR,      State::ERROR,      State::ID,         State::ERROR,      State::ERROR,      State::ERROR,   State::ERROR,   State::PARAM_STAR, State::ERROR,   State::ERROR,      State::ERROR},
        /* ID */        {State::ERROR,      State::ERROR,      State::ERROR,      State::LPAREN,      State::ERROR,      State::ERROR,   State::ERROR,   State::ERROR,      State::ERROR,   State::ERROR,      State::ERROR},
        /* LPAREN */    {State::ERROR,      State::TYPE,       State::ERROR,      State::ERROR,      State::ERROR,      State::ERROR,   State::ERROR,   State::ERROR,      State::ERROR,   State::ERROR,      State::ERROR},
        /* PARAM_LIST */{State::ERROR,      State::ERROR,      State::ERROR,      State::ERROR,      State::RPAREN,     State::ERROR,   State::ERROR,   State::ERROR,      State::TYPE,    State::ERROR,      State::TYPE},  // Исправлено
        /* PARAM_STAR */{State::ERROR,      State::ERROR,      State::ID,         State::ERROR,      State::ERROR,      State::ERROR,   State::ERROR,   State::PARAM_STAR, State::ERROR,   State::ERROR,      State::ERROR},
        /* ARRAY_DIM */  {State::ERROR,      State::ERROR,      State::ERROR,      State::ERROR,      State::ERROR,      State::NUMBER,  State::PARAM_LIST,  State::ERROR,      State::ERROR,   State::ERROR,      State::ERROR}, // Исправлено
        /* NUMBER */    {State::ERROR,      State::ERROR,      State::ERROR,      State::ERROR,      State::ERROR,      State::ERROR,   State::ARRAY_DIM,    State::ERROR,      State::ERROR,   State::ERROR,      State::ERROR},
        /* RPAREN */    {State::ERROR,      State::ERROR,      State::ERROR,      State::ERROR,      State::ERROR,      State::ERROR,   State::ERROR,   State::ERROR,      State::ERROR,   State::END,       State::ERROR},
        /* END */       {State::ERROR,      State::ERROR,      State::ERROR,      State::ERROR,      State::ERROR,      State::ERROR,   State::ERROR,   State::ERROR,      State::ERROR,   State::ERROR,      State::ERROR},
};


Lexem::Type getLexemType(const std::string& token) {
    if (token == "void" || token == "int" || token == "double" || token == "float" ||
        token == "char" || token == "bool" || token=="long") {
        return Lexem::TYPE;
    } else if (isalpha(token[0]) || token[0] == '_') {
        for (size_t i = 1; i < token.size(); ++i) {
            if (!isalnum(token[i]) && token[i] != '_') {
                return Lexem::UNKNOWN;
            }
        }
        return Lexem::ID;
    } else if (token == "(") {
        return Lexem::LPAREN;
    } else if (token == ")") {
        return Lexem::RPAREN;
    } else if (token == "[") {
        return Lexem::LBRACKET;
    } else if (token == "]") {
        return Lexem::RBRACKET;
    } else if (token == "*") {
        return Lexem::STAR;
    } else if (token == ",") {
        return Lexem::COMMA;
    } else if (token == ";") {
        return Lexem::SEMICOLON;
    } else if (isdigit(token[0])) {
        for (size_t i = 1; i < token.size(); ++i) {
            if (!isdigit(token[i])) {
                return Lexem::UNKNOWN;
            }
        }
        return Lexem::NUMBER;
    } else {
        return Lexem::UNKNOWN;
    }
}

std::vector<std::string> tokenize(const std::string& input) {
    std::vector<std::string> tokens;
    std::string currentToken;

    for (size_t i = 0; i < input.size(); ++i) {
        char c = input[i];

        if (isspace(c) || c == '(' || c == ')' || c == '[' || c == ']' || c == ',' || c == ';') {
            if (!currentToken.empty()) {
                tokens.push_back(currentToken);
                currentToken.clear();
            }
            if (!isspace(c)) {
                tokens.push_back(std::string(1, c));
            }
        } else if (c == '*') { // Обработка '*'
            if (!currentToken.empty()) { // Если перед '*' есть символы, добавляем их как токен
                tokens.push_back(currentToken);
                currentToken.clear();
            }
            tokens.push_back("*"); // Добавляем '*' как отдельный токен
        } else {
            currentToken += c;
        }
    }

    if (!currentToken.empty()) {
        tokens.push_back(currentToken);
    }

    std::cout << "TOKENS:\n";
    for (auto& t: tokens) {
        std::cout << t << std::endl;
    }
    return tokens;
}

void printResults(const std::string& functionType,
                  const std::string& functionName,
                  const std::vector<std::pair<std::string, std::pair<std::string, std::vector<int>>>>& parameters) {
    std::cout << "Function type: " << functionType << std::endl;
    std::cout << "Function name: " << functionName << std::endl;
    std::cout << "Parameters: " << std::endl;
    for (const auto& param : parameters) {
        std::cout << "  " << param.first << " "
                  << param.second.first << ": ";
        for (int dim : param.second.second) {
            std::cout << "[" << dim << "]";
        }
        std::cout << std::endl;
    }
}

bool analyze(const std::vector<std::string>& tokens) {
    State::Type currentState = State::START;
    std::string functionType;
    std::string functionName;
    std::vector<std::pair<std::string, std::pair<std::string, std::vector<int>>>> parameters;
    std::pair<std::string, std::pair<std::string, std::vector<int>>> currentParam;

    for (const std::string& token : tokens) {
        Lexem::Type lexemType = getLexemType(token);
        currentState = transitionTable[currentState][lexemType];

        switch (currentState) {
            case State::TYPE:
                if (functionType.empty())
                    functionType = token;
                else
                    currentParam.first = token; // Тип параметра
                break;
            case State::ID:
                if (functionName.empty()) {
                    functionName = token;
                } else {
                    currentParam.second.first = token;
                }
                break;
            case State::PARAM_STAR: // Обработка указателей
                currentParam.first += "*";
                break;
            case State::NUMBER:
                currentParam.second.second.push_back(std::stoi(token));
                break;
            case State::PARAM_LIST:
                if (!currentParam.second.first.empty()){
                    parameters.push_back(currentParam);
                }
                currentParam = {};
                break;
            case State::ERROR:
                if (!currentParam.second.first.empty()){
                    parameters.push_back(currentParam);
                }
                printResults(functionType, functionName, parameters);
                std::cout << "Error: Invalid syntax." << std::endl;
                return false;
            case State::END:
                if (!currentParam.second.first.empty()){
                    parameters.push_back(currentParam);
                }
                printResults(functionType, functionName, parameters);
                return true;
            default:
                break;
        }
    }

    return false;
}

int main() {
    std::string input;
    std::cout << "Enter procedure declaration: ";
    std::getline(std::cin, input);

    std::vector<std::string> tokens = tokenize(input);
    analyze(tokens);

    std::cout << getLexemType("**kk");
    return 0;
}

File: ASM/hw/hw2/main.py
import re

# Классы токенов
TOKEN_TYPES = {
    'TYPE': r'\b(?:void|int|float|long|char|double)\b',  # Типы данных
    'ID': r'[a-zA-Z_][a-zA-Z0-9_]*',  # Идентификатор
    'NUMBER': r'\d+',  # Число
    'OP': r'[\+\-]',  # Операторы сложения и вычитания
    'STAR': r'\*+',  # Одна или несколько звездочек
    'LPAREN': r'\(',  # Левая скобка
    'RPAREN': r'\)',  # Правая скобка
    'LBRACKET': r'\[',  # Левая квадратная скобка
    'RBRACKET': r'\]',  # Правая квадратная скобка
    'COMMA': r',',  # Запятая
    'SEMICOLON': r';'  # Точка с запятой
}


class Token:
    def __init__(self, type, value):
        self.type = type
        self.value = value

    def __repr__(self):
        return f'Token({self.type}, {repr(self.value)})'


def lex(input_string):
    tokens = []
    position = 0
    while position < len(input_string):
        match = None
        if input_string[position].isspace():
            position += 1
            continue
        for token_type, token_regex in TOKEN_TYPES.items():
            pattern = re.compile(token_regex)
            match = pattern.match(input_string, position)
            if match:
                token = Token(token_type, match.group(0))
                tokens.append(token)
                position = match.end(0)
                break
        if not match:
            raise SyntaxError(f'Unexpected character: {input_string[position]}')
    return tokens


class Parser:
    def __init__(self, tokens):
        self.tokens = tokens
        self.position = 0

    def parse(self):
        return self.function_declaration()

    def function_declaration(self):
        self.data_type()
        self.identifier()
        self.eat('LPAREN')
        self.parameter_list()
        self.eat('RPAREN')
        self.eat('SEMICOLON')
        return "Function declaration parsed successfully"

    def data_type(self):
        if self.current_token().type == 'TYPE':
            self.eat('TYPE')
        else:
            raise SyntaxError(f'Expected data type, found {self.current_token()}')

    def identifier(self):
        if self.current_token().type == 'ID':
            self.eat('ID')
        else:
            raise SyntaxError(f'Expected identifier, found {self.current_token()}')

    def parameter_list(self):
        if self.current_token().type == 'TYPE':
            self.parameter()
            while self.current_token().type == 'COMMA':
                self.eat('COMMA')
                self.parameter()

    def parameter(self):
        self.data_type()
        while self.current_token().type == 'STAR':
            self.eat('STAR')
        self.identifier()
        while self.current_token().type == 'LBRACKET':
            self.eat('LBRACKET')
            self.eat('NUMBER')
            self.eat('RBRACKET')

    def current_token(self):
        return self.tokens[self.position]

    def eat(self, token_type):
        if self.current_token().type == token_type:
            self.position += 1
        else:
            raise SyntaxError(f'Unexpected token: {self.current_token()}')

    def pretty_print_declaration(self):
        function_name = ''
        return_type = ''
        parameters = []

        i = 0
        while i < len(self.tokens):
            if tokens[i].type == 'TYPE':
                return_type = tokens[i].value
            elif tokens[i].type == 'ID' and not function_name:
                function_name = tokens[i].value
            elif tokens[i].type == 'LPAREN':
                i += 1
                while tokens[i].type != 'RPAREN':
                    parameter_type = ''
                    pointers = ''
                    arrays = []

                    while tokens[i].type == 'TYPE':
                        parameter_type += tokens[i].value + ' '
                        i += 1

                    while tokens[i].type == 'STAR':
                        pointers += tokens[i].value
                        i += 1

                    parameter_name = tokens[i].value
                    i += 1

                    while tokens[i].type == 'LBRACKET':
                        arrays.append(tokens[i + 1].value)
                        i += 2

                    parameters.append({
                        'type': parameter_type,
                        'name': parameter_name,
                        'pointers': pointers,
                        'arrays': arrays
                    })

                    if tokens[i].type == 'COMMA':
                        i += 1

            i += 1

        print()
        print(f"{return_type} {function_name}")
        print("параметры:")
        for parameter in parameters:
            if parameter['name'] != ']':
                print()
                print(f"тип: {parameter['type']}")
                print(f"идентификатор: {parameter['name']}")
                print(f"казатели: {parameter['pointers']}")
                print(f"квадратные скобки: {', '.join(parameter['arrays'])}")
                print()


input_string = input("Введите выражение: ")
tokens = lex(input_string)
print("\nTokens:", tokens)

parser = Parser(tokens)
try:
    result = parser.parse()
    parser.pretty_print_declaration()

    print("OK")
except SyntaxError as e:
    print("ERROR")
    print(e)
